functionalDB/chapter.md

```diff
@@ -0,0 +1,1138 @@
+   (drop-entity [storage entity] (dissoc storage (:id entity))))
+````
+
+### Querying our data
+
+Now that we've defined the basic elements of our database, we can start thinking about how we're going to query it. By virtue of how we've structured our data, any query is necessarily going to be interested in at least one of an entity's id, and the name and value of some of its attributes. This triplet of (entity-id, attribute-name, attribute-value) is important enough to our query process that we give it an explicit name -- a _datom_.
+
+The reason that datoms are so important is that they represent facts, and our database accumulates facts. 
+
+If you've used a database system before, you are probably already familiar with the concept of an _index_, which is a supporting data structure that consumes extra space in order to decrease the average query time.  In our database, an index is a three-leveled structure, which stores the components of a datom in a specific order. Each index derives its name from the order it stores the datom's components.
+
+For example, let’s look at at the index sketched in Figure 2:
+* the first level stores entity-ids (the blue-ish area) 
+* the second level stores the related attribute-names (the green-ish area)
+* the third level stores the related value (the pink-ish area)
+
+This index is named EAVT, as the top level map holds (E) entity ids, the second level holds (A) attribute names, and the leaves hold (V) values. The (T) comes from the fact that each layer in the database has its own indexes, hence the index itself is relevant for a specific (T) time. 
+
+![image alt text](image_1.png)
+
+Figure 2
+
+Figure 3 shows an index that would be called AVET since:
+
+* First level map holds attribute-name
+* Second level map holds the values (of the attributes)
+* Third level set holds the entity-ids (of the entities whose attribute is at the first level) 
+
+![image alt text](image_2.png)
+
+Figure 3
+
+Our indexes are implemented as a map of maps, where the keys of the root map act as the first level, each such key points to a map whose keys act as the index’s second-level and the values are the index’s third level. Each element in the third level is a set, holding the leaves of the index.
+
+Each index stores the components of a datom as some permutation of its canonical 'EAV' ordering (entity_id, attribute-name, attribute-value). However, when we are working with datoms _outside_ of the index, we expect them to be in canonical format. We thus provide each index with functions *from-eav* and *to-eav* to convert to and from these orderings.
+
+In most database systems, indexes are an optional component; for example, in an RDMBS like postgresql or mysql, you will choose to add indexes only to certain columns in a table. We provide each index with a *usage-pred* function that determines for an attribute whether it should be included in this index or not. 

```

Comment:
RDMBS -> RDBMS? Might be worth defining the acronym, too, since it's the first time we've seen it. Or dropping it altogether since it's the only time it's used in the chapter.